From 43b298f7ce076eec97ef0ae4f8bb6aac73367c8f Mon Sep 17 00:00:00 2001
From: "Thulasimani,Sivakumar" <sivakumar.thulasimani@intel.com>
Date: Sat, 10 Oct 2015 01:10:53 +0530
Subject: [PATCH 17/28] FOR_UPSTREAM [VPG]: drm/i915: Fix EQ sequence in Link
 training

Second half of link training for DP panel is check for equalization.
it is expected that CR should always pass when checking for EQ,
hence this patch fixes this sequence by exiting from function
where EQ is checked if CR fails and returns failure to indicate
the caller of the EQ status.

Change-Id: I039f4a582ab2001cfc92bf20d0779bba36f2d521
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-9749
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
Reviewed-on: https://android.intel.com:443/439690
---
 drivers/gpu/drm/i915/intel_dp.c  | 13 +++++--------
 drivers/gpu/drm/i915/intel_drv.h |  2 +-
 2 files changed, 6 insertions(+), 9 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_dp.c b/drivers/gpu/drm/i915/intel_dp.c
index 8e90d98..c01fa5f 100644
--- a/drivers/gpu/drm/i915/intel_dp.c
+++ b/drivers/gpu/drm/i915/intel_dp.c
@@ -3961,7 +3961,7 @@ intel_dp_start_link_train(struct intel_dp *intel_dp)
 	return ret;
 }
 
-void
+bool
 intel_dp_complete_link_train(struct intel_dp *intel_dp)
 {
 	bool channel_eq = false;
@@ -3978,7 +3978,7 @@ intel_dp_complete_link_train(struct intel_dp *intel_dp)
 				     training_pattern |
 				     DP_LINK_SCRAMBLING_DISABLE)) {
 		DRM_ERROR("failed to start channel equalization\n");
-		return;
+		return channel_eq;
 	}
 
 	tries = 0;
@@ -4001,12 +4001,8 @@ intel_dp_complete_link_train(struct intel_dp *intel_dp)
 
 		/* Make sure clock is still ok */
 		if (!drm_dp_clock_recovery_ok(link_status, intel_dp->lane_count)) {
-			intel_dp_start_link_train(intel_dp);
-			intel_dp_set_link_train(intel_dp, &DP,
-						training_pattern |
-						DP_LINK_SCRAMBLING_DISABLE);
-			++tries;
-			continue;
+			DRM_ERROR("CR Failed inside EQ check\n");
+			break;
 		}
 
 		if (drm_dp_channel_eq_ok(link_status, intel_dp->lane_count)) {
@@ -4031,6 +4027,7 @@ intel_dp_complete_link_train(struct intel_dp *intel_dp)
 		DRM_DEBUG_KMS("Channel EQ done. DP Training successful\n");
 	}
 
+	return channel_eq;
 }
 
 void intel_dp_stop_link_train(struct intel_dp *intel_dp)
diff --git a/drivers/gpu/drm/i915/intel_drv.h b/drivers/gpu/drm/i915/intel_drv.h
index d061a37..54caf94 100644
--- a/drivers/gpu/drm/i915/intel_drv.h
+++ b/drivers/gpu/drm/i915/intel_drv.h
@@ -998,7 +998,7 @@ void intel_dp_init(struct drm_device *dev, int output_reg, enum port port);
 bool intel_dp_init_connector(struct intel_digital_port *intel_dig_port,
 			     struct intel_connector *intel_connector);
 bool intel_dp_start_link_train(struct intel_dp *intel_dp);
-void intel_dp_complete_link_train(struct intel_dp *intel_dp);
+bool intel_dp_complete_link_train(struct intel_dp *intel_dp);
 void intel_dp_stop_link_train(struct intel_dp *intel_dp);
 bool intel_dp_fast_link_train(struct intel_dp *intel_dp);
 bool chv_upfront_link_train(struct drm_device *dev,
-- 
1.9.1

